src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/JsonEncoders/CompressedMessageJsonEncoder.cs

```diff
@@ -0,0 +1,166 @@
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Driver.Core.Compression;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.WireProtocol.Messages.Encoders.BinaryEncoders;
+
+namespace MongoDB.Driver.Core.WireProtocol.Messages.Encoders.JsonEncoders
+{
+    /// <summary>
+    /// Represents a Json encoder for a CompressedMessage.
+    /// </summary>
+    /// <seealso cref="MongoDB.Driver.Core.WireProtocol.Messages.Encoders.IMessageEncoder" />
+    public class CompressedMessageJsonEncoder : MessageJsonEncoderBase, IMessageEncoder
+    {
+        private readonly MessageEncoderSettings _encoderSettings;
+
+        // constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompressedMessageJsonEncoder"/> class.
+        /// </summary>
+        /// <param name="textReader">The text reader.</param>
+        /// <param name="textWriter">The text writer.</param>
+        /// <param name="encoderSettings">The encoder settings.</param>
+        public CompressedMessageJsonEncoder(TextReader textReader, TextWriter textWriter, MessageEncoderSettings encoderSettings) : base(textReader, textWriter, encoderSettings)
+        {
+            _encoderSettings = encoderSettings;
+        }
+
+        // public methods
+        /// <summary>
+        /// Reads the message.
+        /// </summary>
+        /// <returns>A message.</returns>
+        public CompressedMessage ReadMessage()
+        {
+            var reader = CreateJsonReader();
+            var context = BsonDeserializationContext.CreateRoot(reader);
+            var messageDocument = BsonDocumentSerializer.Instance.Deserialize(context);
+
+            var opcode = (Opcode)messageDocument["opcode"].ToInt32();
+            if (opcode != Opcode.Compressed)
+            {
+                throw new FormatException($"Command message invalid opcode: \"{opcode}\".");
+            }
+
+            var originalOpcode = (Opcode)messageDocument.GetValue("originalOpcode").ToInt32();
+            var compressorId = (CompressorType)messageDocument["compressorId"].ToInt32();
+            var compressedMessage = messageDocument["compressedMessage"].AsString;
+
+            using (var textReader = new StringReader(compressedMessage))
+            {
+                var originalEncoder = GetEncoderByOpcode(opcode: originalOpcode, textReader: textReader);
+                var originalMessage = originalEncoder.ReadMessage();
+
+                return new CompressedMessage(originalMessage, null, compressorId);
```

Comment:
I wrote above, about replacing the `null` parameter on the `JsonEncoder` implementation for the new possible interface: `ICompressedMessageContent `